src/StackExchange.Redis/CommandRetry/MessageRetryQueue.cs

```diff
@@ -0,0 +1,172 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+
+namespace StackExchange.Redis
+{
+    internal class MessageRetryQueue : IDisposable
+    {
+        private readonly Queue<Message> _queue = new Queue<Message>();
+        private readonly IMessageRetryHelper _messageRetryHelper;
+        private readonly int? _maxRetryQueueLength;
+        private readonly bool _runRetryLoopAsync;
+
+        internal MessageRetryQueue(IMessageRetryHelper messageRetryHelper, int? maxRetryQueueLength = null, bool runRetryLoopAsync = true)
```

Comment:
@deepakverma Trying to figure out the scenario in which `runRetryLoopAsync` would be `false` (triggering a `.Wait()` below), can you shed more light on that scenario/path?